### PR TITLE
[MM-33083] Use HTTP request like payload to invoke lambda

### PR DESCRIPTION
**Pull request by hanzei**
#### Summary
Because the Apps expect HTTP-like requests we have to adopt the payload that is used to invoke the lambda function. The format is documented [here](https://docs.aws.amazon.com/apigateway/latest/developerguide/http-api-develop-integrations-lambda.html). I didn't fill out all fields given that we only need a handful.

There is some double JSON Unmarshaling in the code that I couldn't prevent. Maybe we need to optimize that later.

Given that there is no per path mapping to function right now, all calls go to the first defined function. Hence, the PR kind of drops support for more then one function.

I've tested the PR with the Zendesk app and can confirm that it works.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-33083

### CODE REVIEW

**Review by levb — APPROVED** [review]

**Comment by levb** [comment]
Thank you, this should make the app developer's job easier.

File: server/api/impl/upstream/upawslambda/upstream.go [hunk 1]
```diff
@@ -27,14 +42,52 @@ func NewUpstream(app *apps.App, aws *aws.Client) *Upstream {
 }
 
 func (u *Upstream) OneWay(call *apps.Call) error {
-	_, err := u.aws.InvokeLambda(u.app.Manifest.AppID, u.app.Manifest.Version, call.URL, lambda.InvocationTypeEvent, call)
+	payload, err := CallToInvocationPayload(call)
+	if err != nil {
+		return errors.Wrap(err, "failed to covert call into invocation payload")
+	}
+
+	_, err = u.aws.InvokeLambda(u.app.Manifest.AppID, u.app.Manifest.Version, u.app.Manifest.Functions[0].Name, lambda.InvocationTypeEvent, payload)
```
Comment (iomodo) [inline server/api/impl/upstream/upawslambda/upstream.go:50]: why do we invoke `u.app.Manifest.Functions[0].Name`?

File: server/api/impl/upstream/upawslambda/upstream.go [hunk 2]
```diff
@@ -27,14 +42,52 @@ func NewUpstream(app *apps.App, aws *aws.Client) *Upstream {
...
+
+	if resp.StatusCode != http.StatusOK {
+		return nil, errors.Errorf("lambda invocation failed with status code %v and body %v", resp.StatusCode, resp.Body)
+	}
+
+	return ioutil.NopCloser(strings.NewReader(resp.Body)), nil
+}
+
+func CallToInvocationPayload(call *apps.Call) ([]byte, error) {
+	request := invocationPayload{
```
Comment (iomodo) [inline server/api/impl/upstream/upawslambda/upstream.go:80]: What is the reason of having the HTTP like payload? Matching with the HTTP apps?

File: server/api/impl/upstream/upawslambda/upstream.go (hunk 1 above)
Comment (hanzei) [reply server/api/impl/upstream/upawslambda/upstream.go:50]: What mechanism do we have to decide which function to invoke?

Comment (iomodo) [reply server/api/impl/upstream/upawslambda/upstream.go:50]: It's in the `call`

Comment (hanzei) [reply server/api/impl/upstream/upawslambda/upstream.go:50]: Maybe I'm misunderstanding. What field in `Call` can be used to to decide which function to invoke? A Call might only have a `Path` which we can only use for routing if a function also have something like a `Path`.

File: server/api/impl/upstream/upawslambda/upstream.go (hunk 2 above)
Comment (hanzei) [reply server/api/impl/upstream/upawslambda/upstream.go:80]: Apps expect an HTTP like payload given that they are essentially HTTP servers. That approach makes the transition from an HTTP app to an AWS app seamless. Given that we don't place an AWS Gateway in front of the lambda we have to generate the payload ourself.

Comment (levb) [reply server/api/impl/upstream/upawslambda/upstream.go:80]: I 👍  @hanzei's view here. This approach would also make a transition to a self-hosted (lambda) easier on the developer, should they choose to do so.

**Review by iomodo — COMMENTED** [review]
My only concern with this PR is `u.app.Manifest.Functions[0].Name`
To map `path` to the functions name, we could assume that all path formats are `func_name/somepath...` @hanzei thoughts?

**Comment by hanzei** [comment]
@iomodo I don't think that assumption is valid. E.g. `/install` is the default path which would ever be valid then. We've decided offline that we want multi function apps, but I'm hesitant to add support for that in this PR. @levb is working on that anyway.

**Comment by iomodo** [comment]
This PR will work only for a single function. I'm OK merging as is. We can track the multi-function effort in the separate PR

**Review by iomodo — APPROVED** [review]

**Pull request merged**